pkg/webhook/user/mutation/mutation.go

```diff
@@ -0,0 +1,122 @@
+limitations under the License.
+*/
+
+package mutation
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/go-logr/logr"
+
+	kubermaticv1 "k8c.io/kubermatic/v2/pkg/apis/kubermatic/v1"
+	kubermaticv1helper "k8c.io/kubermatic/v2/pkg/apis/kubermatic/v1/helper"
+	"k8c.io/kubermatic/v2/pkg/webhook/util"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	ctrlruntime "sigs.k8s.io/controller-runtime"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// AdmissionHandler for mutating Kubermatic User CRD.
+type AdmissionHandler struct {
+	client  ctrlruntimeclient.Client
+	log     logr.Logger
+	decoder *admission.Decoder
+}
+
+// NewAdmissionHandler returns a new User AdmissionHandler.
+func NewAdmissionHandler(client ctrlruntimeclient.Client) *AdmissionHandler {
+	return &AdmissionHandler{
+		client: client,
+	}
+}
+
+func (h *AdmissionHandler) SetupWebhookWithManager(mgr ctrlruntime.Manager) {
+	mgr.GetWebhookServer().Register("/mutate-kubermatic-k8c-io-v1-user", &webhook.Admission{Handler: h})
+}
+
+func (h *AdmissionHandler) InjectLogger(l logr.Logger) error {
+	h.log = l.WithName("user-mutation-handler")
+	return nil
+}
+
+func (h *AdmissionHandler) InjectDecoder(d *admission.Decoder) error {
+	h.decoder = d
+	return nil
+}
+
+func (h *AdmissionHandler) Handle(ctx context.Context, req webhook.AdmissionRequest) webhook.AdmissionResponse {
+	newUser := &kubermaticv1.User{}
+	oldUser := &kubermaticv1.User{}
+
+	switch req.Operation {
+	case admissionv1.Create:
+		if err := h.decoder.Decode(req, newUser); err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+
+		if err := h.ensureProjectRelation(ctx, newUser, nil); err != nil {
+			return webhook.Errored(http.StatusBadRequest, err)
+		}
+
+	case admissionv1.Update:
+		if err := h.decoder.Decode(req, newUser); err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+		if err := h.decoder.DecodeRaw(req.OldObject, oldUser); err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+
+		if err := h.ensureProjectRelation(ctx, newUser, oldUser); err != nil {
+			return webhook.Errored(http.StatusBadRequest, err)
+		}
+
+	case admissionv1.Delete:
+		return webhook.Allowed(fmt.Sprintf("no mutation done for request %s", req.UID))
+
+	default:
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("%s not supported on user resources", req.Operation))
+	}
+
+	mutatedKey, err := json.Marshal(newUser)
+	if err != nil {
+		return webhook.Errored(http.StatusInternalServerError, fmt.Errorf("marshaling user object failed: %w", err))
+	}
+
+	return admission.PatchResponseFromRaw(req.Object.Raw, mutatedKey)
+}
+
+func (h *AdmissionHandler) ensureProjectRelation(ctx context.Context, user *kubermaticv1.User, oldUser *kubermaticv1.User) error {
```

Comment:
ensureProjectRelation (for both user and sshkeys) is now not mutating anything, but validating. Maybe for the serviceaccounts we dont need the mutating webhook at all and we can move this to the validating webhook ( it already validates that project is immutable, so jus the OptimisticallyCheckIfProjectIsValid

Reply:
Excellent point.

Reply:
In fact, since we do not set the owner ref anymore, there is no need for a mutating webhook for users. So I moved the validation logic into the validating webhook and removed the mutating one.